### PR TITLE
Feature/keyboard unit tests

### PR DESCRIPTION
**Pull request by AlexandreSanscartier**

### CODE REVIEW

File: examples/KeyboardDemo/KeyboardDemoScene.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import KeyboardKey from 'inputEngine/KeyboardKey';
+
+export default class KeyboardDemoScene extends Scene {
+  create() {
+    this.game.Keyboard.addKey(KeyboardKey.A);
+  }
+
+  update() {
+    if (this.game.Keyboard.keyDownHeld(KeyboardKey.A)) {
+      console.log('A is down');
```
Comment (Zargath) [inline examples/KeyboardDemo/KeyboardDemoScene.js]: We could change text on the screen to make this more exciting

File: src/inputEngine/Keyboard.js [hunk 2]
```diff
@@ -1,40 +1,45 @@
...
+  }
 
-    // Fill the array with the possible keyboard keys
-    // TODO: Validate that this is fin
-    this.keyState = new Array(this.KEY_COUNT);
-    for (let i = 0; i < this.KEY_COUNT; i += 1) {
-      this.keyState[i] = this.keyStates.RELEASED;
+  registerKey(key) {
+    if (this.retrieveKey(key) !== undefined) {
+      throw new ItemAlreadyExistsError();
```
Comment (Zargath) [inline src/inputEngine/Keyboard.js]: I don't think an error should be thrown, simply make sure it's added.

File: src/inputEngine/KeyboardKey.js [hunk 3]
```diff
@@ -0,0 +1,892 @@
+// Used to define a KeyboardKey (A) for example
+// Use the KeyEvent.code to find out which physical key on the keyboard was PRESSED
+// Use the KeyEvent.key to find out which key that physical key maps to
+class KeyboardKey {
```
Comment (Zargath) [inline src/inputEngine/KeyboardKey.js]: Missing Documentation

File: tests/Keyboard.test.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+});
+
+describe('retrieveKey', () => {
+  it('should retrieve the right key when using the KeyEvent.code', () => {
+    // Arrange
+    const keyboard = new Keyboard();
+    const code = 'KeyA';
+
+    // Act
+    keyboard.registerKey(KeyboardKey.A);
```
Comment (Zargath) [inline tests/Keyboard.test.js]: Should be in the Arrange

File: tests/Keyboard.test.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    // Assert
+    expect(() => { keyboard.registerKey(KeyboardKey.A); }).toThrow(ItemAlreadyExistsError);
+  });
+});
+
+describe('retrieveKey', () => {
+  it('should retrieve the right key when using the KeyEvent.code', () => {
+    // Arrange
+    const keyboard = new Keyboard();
+    const code = 'KeyA';
```
Comment (Zargath) [inline tests/Keyboard.test.js]:
```suggestion
    const code = KeyboardKey.A.code;
```

File: examples/KeyboardDemo/KeyboardDemoScene.js (hunk 1 above)
Comment (AlexandreSanscartier) [reply examples/KeyboardDemo/KeyboardDemoScene.js]: created a new issue to resolve this.

File: src/inputEngine/Keyboard.js (hunk 2 above)
Comment (AlexandreSanscartier) [reply src/inputEngine/Keyboard.js]: Agreed.

File: tests/Keyboard.test.js (hunk 5 above)
Comment (AlexandreSanscartier) [reply tests/Keyboard.test.js]: Fixed

**Comment by AlexandreSanscartier** [comment]
Fixed the changes I'll be merging the pull request now thanks!

**Pull request merged**